action/protocol/poll/governance_protocol.go

```diff
@@ -412,60 +400,85 @@ func (p *governanceChainCommitteeProtocol) calculateKickoutBlackList(
 				break
 			}
 			// N-1, N-2, ..., N-K
-			numBlks, produce, err := p.productivityByEpoch(ctx, epochNum-round)
+			isCurrentEpoch := false
+			if round == 1 {
+				isCurrentEpoch = true
+			}
+			uq, err := p.calculateUnproductiveDelegatesByEpoch(ctx, epochNum-round, isCurrentEpoch)
 			if err != nil {
 				return nil, err
 			}
-			if round == 1 {
-				// The current block is not included, so that we need to add it to the stats
-				numBlks++
-				produce[blkCtx.Producer.String()]++
-			}
-			expectedNumBlks := numBlks / uint64(len(produce))
-			for addr, actualNumBlks := range produce {
-				if actualNumBlks*100/expectedNumBlks < p.productivityThreshold {
-					unqualifiedDelegates = append(unqualifiedDelegates, addr)
+			for _, addr := range uq {
+				if _, ok := unqualifiedDelegates[addr]; !ok {
+					unqualifiedDelegates[addr] = 1
+				} else {
+					unqualifiedDelegates[addr]++
 				}
 			}
 			round--
 		}
 		return unqualifiedDelegates, nil
 	}
 
-	// Blacklist[N] = Blacklist[N-1] - BlackList[N-K] + Low-productivity-list[N-1]
-	prevBlacklist, err := p.kickoutListByEpoch(p.sr, epochNum-1)
+	// Blacklist[N] = Blacklist[N-1] - Low-productivity-list[N-K-1] + Low-productivity-list[N-1]
+	blackListMapping, err := p.kickoutListByEpoch(p.sr, epochNum-1)
+	if err != nil {
+		return nil, err
+	}
+	skipList, err := p.calculateUnproductiveDelegatesByEpoch(ctx, epochNum-p.kickoutEpochPeriod-1, false)
```

Comment:
if epochNum <= p.kickoutEpochPeriod, something may go wrong.

Reply:
line 395, handle ```if epochNum <= englishEpochNum+p.kickoutEpochPeriod ```case 
so, epochNum is always > englishEpochNum+p.kickoutEpochPeriod